libazureinit/src/provision/mod.rs

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+pub mod hostname;
+pub mod password;
+pub(crate) mod ssh;
+pub mod user;
+
+use strum::IntoEnumIterator;
+use tracing::instrument;
+
+use crate::{error::Error, imds::PublicKeys};
+
+/// The interface for applying the desired configuration to the host.
+///
+/// Optional settings, like a password for the user account, can be provided
+/// after constructing this object. By default, all known methods for
+/// provisioning a particular setting are tried until one succeeds. Particular
+/// methods can be selected via the `*_provisioners()` methods
+/// ([`Provision::hostname_provisioners`], [`Provision::user_provisioners`],
+/// etc).
+///
+/// To actually apply the configuration, use [`Provision::provision`].
+#[derive(Default, Clone)]
+pub struct Provision {
+    hostname: String,
+    username: String,
+    keys: Vec<PublicKeys>,
+    password: Option<String>,
+    hostname_backends: Option<Vec<hostname::Provisioner>>,
+    user_backends: Option<Vec<user::Provisioner>>,
+    password_backends: Option<Vec<password::Provisioner>>,
+}
+
+impl Provision {
+    pub fn new(
+        hostname: impl Into<String>,
+        username: impl Into<String>,
+        ssh_keys: impl Into<Vec<PublicKeys>>,
+    ) -> Self {
```

Comment:
The code looks okay as long as we have only 3 parameters for `new()`.

Just a thought. I am wondering how much scalable it could be, for example when we would have 10 - 20 backends.
Then callers need to pass 10 - 20 parameters to `new()`, which could be tricky to follow and maintain.
However, as far as I understand, Rust does not have a native way of dealing variable args, right?
So we might need to split those into separate structs in the future.

Reply:
Yeah, I think `new()` is just for required arguments and any change there would be an API breakage. I felt like these three would always be required, but we could also accept no arguments at all to `new()` and if you don't want to create a user for whatever reason, you don't have to. At that point, though, I don't know why you'd use the API.

Reply:
When adding the group I added a dedicated type for user configuration so it reduces the number of parameters to two.